Address Output Report - System Sizing Information Time of Peak Cooling and User Design Capacity In IP Units are Wrong

Nigusse opened the pull request:
Pull request overview
---------------------
 - Fixes #9952 
 - Corrects the system sizing peak cooling `Date/Time of Peak` column of the `System Sizing Information` tabular and the eio output files. Previously, the `Time of Peak` information was incorrect. Warning error message is expected in several example files because of the `System Sizing Information` table `Date/Time of Peak` column value change for cooling.
 - Added missing unit for `User Design Capacity` header in `System Sizing Information` table. `EIO` and `html` output tables header difference is expected. Large `html` table and `eio` output diffs are expected if `IP Units` is specified for `OutputControl:Table:Style`.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

Nigusse committed:
fixes the system sizing peak cooling load timestamp

Nigusse committed:
Modifying an existing unit test to demonstrate the fix

Nigusse committed:
Merge branch 'develop' into Issue9952_SystemSizingCoolPeak

Nigusse commented:
Diffs are expected for several regression test suite files (300+) because of change on the system sizing peak cooling `Date/Time of Peak` column of the `System Sizing Information` `html` tabular and the `eio` output files. Actually the date is correct, it is the time, `hh:mm:ss` that will changed. Sample diffs from test suite eio file:

```
! , System Name, Load Type, Peak Load Kind, User Design Capacity, Calc Des Air Flow Rate [m3/s], User Des Air Flow Rate [m3/s], Design Day Name, Date/Time of Peak
- System Sizing Information, VAV SYS 1, Cooling, Sensible, 21030.82, 1.55096, 1.55096, CHICAGO ANN CLG 1% CONDNS DB=>MWB, 7/21 00:00:00
+ System Sizing Information, VAV SYS 1, Cooling, Sensible, 21030.82, 1.55096, 1.55096, CHICAGO ANN CLG 1% CONDNS DB=>MWB, 7/21 15:45:00
```

Nigusse commented:
System Sizing `Date/Time of Peak` for cooling has changed from `8/21 05:00:00` to `8/21 14:30:00`as shown below based on the defect file:
Develop:
![Develop_System_Sizing_Info](https://user-images.githubusercontent.com/5805126/232782688-c3d91daf-95a7-43af-a287-d1c0a6732dbc.png)
Branch:
![Branch_System_Sizing_Info](https://user-images.githubusercontent.com/5805126/232782925-8c66dc4f-5e2c-4030-aaab-3f2090c45d93.png)
Now `Date/Time of Peak` of this table match the time of peak in `eplusssz.csv` and `epluszsz.csv` files.

Nigusse commented:
Uploaded updated defect file on `../DefectFiles/9000s/9952/` on github.

Nigusse commented on src/EnergyPlus/SizingManager.cc:
```diff
@@ -805,6 +807,7 @@ void ManageSizing(EnergyPlusData &state)
                 coolPeakDD = SysSizPeakDDNum(AirLoopNum).TotCoolPeakDD;
                 coolCap = FinalSysSizing(AirLoopNum).TotCoolCap;
+                timeStepIndexAtPeakCoolLoad = SysSizPeakDDNum(AirLoopNum).TimeStepAtTotCoolPk(coolPeakDD);
             }
```
Defined a local variable `timeStepIndexAtPeakCoolLoad` that holds the index to the system cooling `peak time` depending on the `Type of Load to Size On` user input.

Nigusse commented on src/EnergyPlus/SizingManager.cc:
```diff
@@ -816,7 +819,7 @@ void ManageSizing(EnergyPlusData &state)
                                 FinalSysSizing(AirLoopNum).CoolDesDay,
                                 coolPeakDDDate,
-                                SysSizPeakDDNum(AirLoopNum).TimeStepAtHeatPk(coolPeakDD));
+                                timeStepIndexAtPeakCoolLoad);
```
This is the offending code that needed to change. It should point to `TimeStepAtTotCoolPk(coolPeakDD)`, instead of `TimeStepAtHeatPk(coolPeakDD)`.

Myoldmopar replied:
This makes good sense.

rraustad replied:
Yes, I agree with this change.

Nigusse commented on tst/EnergyPlus/unit/DesiccantDehumidifiers.unit.cc:
```diff
@@ -2888,6 +2888,19 @@ TEST_F(EnergyPlusFixture, DesiccantDehum_OnOASystemTest)
+    EXPECT_EQ(coolPeakDDDate, "7/21");
+    std::string dateHrMin = coolPeakDDDate + " " + SizingManager::TimeIndexToHrMinString(*state, timeStepIndexAtPeakCoolLoad);
+    EXPECT_EQ(dateHrMin, "7/21 10:30:00");
 }
```
Modified an existing code to demonstrate that the correct system sizing information  is reported to the `Date/Time of Peak` column.

Nigusse reviewed (commented):
Walk-through the code changes: now the correct cooling system sizing peak `time index` is passed for reporting.

rraustad commented:
@Nigusse issue #9952 states "In the output report, the System Sizing Information, Peak Load User Design Capacity is wrong. I am unable to find these values anywhere else in the report." Now that the time stamp is fixed, can you look at your results and comment on whether the sizing data is wrong?

Nigusse commented:
> @Nigusse issue #9952 states "In the output report, the System Sizing Information, Peak Load User Design Capacity is wrong. I am unable to find these values anywhere else in the report." Now that the time stamp is fixed, can you look at your results and comment on whether the sizing data is wrong?

I was thinking to send the user some clarification. The cooling capacity and flow rate values were correct before and after the fix, the only problem I was able to confirm and fix is the `Date/Time of Peak` of system cooling sizing.

Nigusse committed:
Linux complained when comparing enum

Nigusse committed:
Merge branch 'develop' into Issue9952_SystemSizingCoolPeak

Nigusse commented:
The user's concern or characterization the issue #9952 can be explained as follows:
 
Sample eio output for AirLoopHVAC FCU-33 from the defect file is shown below. The system sizing `User Design Capacity, W` is `21982.54` and is different from the `Design Size Design Coil Load [W], 22747.84` of the water coil in the air-loop:

```
System Sizing Information, FCU-33, Cooling, Total, 21982.54, 0.92851, 0.92851, SAN ANTONIO INTL AP ANN CLG .4% CONDNS WB=>MDB, 8/21 14:30:00
Component Sizing Information, Coil:Cooling:Water, COIL COOLING WATER 11, Design Size Design Inlet Air Temperature [C], 26.14101
Component Sizing Information, Coil:Cooling:Water, COIL COOLING WATER 11, Design Size Design Inlet Water Temperature [C], 7.22000
Component Sizing Information, Coil:Cooling:Water, COIL COOLING WATER 11, Design Size Design Outlet Air Temperature [C], 10.41879
Component Sizing Information, Coil:Cooling:Water, COIL COOLING WATER 11, Design Size Design Inlet Air Humidity Ratio [kgWater/kgDryAir], 9.89899E-003
Component Sizing Information, Coil:Cooling:Water, COIL COOLING WATER 11, Design Size Design Outlet Air Humidity Ratio [kgWater/kgDryAir], 7.97000E-003
Component Sizing Information, Coil:Cooling:Water, COIL COOLING WATER 11, Design Size Design Coil Load [W], 22747.84572
Component Sizing Information, Coil:Cooling:Water, COIL COOLING WATER 11, Design Size Design Water Flow Rate [m3/s], 8.12500E-004
Component Sizing Information, Coil:Cooling:Water, COIL COOLING WATER 11, Design Size Design Air Flow Rate [m3/s], 0.92851
Water Cooling Coil Capacity Information,Coil:Cooling:Water,COIL COOLING WATER 11,27329.83,17800.37,9529.47,0.65,3289.87,33.37
```

The system sizing and water coil design cooling capacities can be different for the following reasons:
- If the system sizing object air outlet conditions and the water coil design air outlet conditions are different, as is the case in this defect file, the cooling capacities can be different. The system sizing uses the  `Central Cooling Design Supply Air Temperature {C}` and `Central Cooling Design Supply Air Humidity Ratio {kgWater/kgDryAir}` specified in the `Sizing:System` and the `Coil:Cooling:Water` object has its own `Design Outlet Air Temperature` and `Design Outlet Air Humidity Ratio` inputs and these fields were `AutoSized` for this defect file. This explains why the system sizing cooling capacity is `21982.54W` and the cooling coil capacity for FCU-33 is `22747.85W` as shown above.
- In the `Water Cooling Coil Capacity Information` the `User Design Capacity` is `27329.83W`, which is different from the `Component Sizing Information` output of `Design Size Design Coil Load [W], 22747.85` because the capacity in `Water Cooling Coil Capacity Information` is reported at the rated condition, not at the design condition. It is an adjusted value for rated condition and is expected to be different.

Setting the water coil design air outlet conditions the same as that of the `Sizing:System` object, produces matching cooling capacity between `System Sizing Information` and `Component Sizing Information` in the `html` and `eio` output files. Sample eio output for AirLoopHVAC `FCU-33` from the defect file after matching system sizing and water coil design air outlet condition is shown below:

```
System Sizing Information, FCU-33, Cooling, Total, 21982.54, 0.92851, 0.92851, SAN ANTONIO INTL AP ANN CLG .4% CONDNS WB=>MDB, 8/21 14:30:00
Component Sizing Information, Coil:Cooling:Water, COIL COOLING WATER 11, Design Size Design Inlet Air Temperature [C], 26.14101
Component Sizing Information, Coil:Cooling:Water, COIL COOLING WATER 11, Design Size Design Inlet Water Temperature [C], 7.22000
Component Sizing Information, Coil:Cooling:Water, COIL COOLING WATER 11, User-Specified Design Outlet Air Temperature [C], 11.11000
Component Sizing Information, Coil:Cooling:Water, COIL COOLING WATER 11, Design Size Design Inlet Air Humidity Ratio [kgWater/kgDryAir], 9.89899E-003
Component Sizing Information, Coil:Cooling:Water, COIL COOLING WATER 11, User-Specified Design Outlet Air Humidity Ratio [kgWater/kgDryAir], 7.97000E-003
Component Sizing Information, Coil:Cooling:Water, COIL COOLING WATER 11, Design Size Design Coil Load [W], 21982.53802
Component Sizing Information, Coil:Cooling:Water, COIL COOLING WATER 11, Design Size Design Water Flow Rate [m3/s], 7.85165E-004
Component Sizing Information, Coil:Cooling:Water, COIL COOLING WATER 11, Design Size Design Air Flow Rate [m3/s], 0.92851
Water Cooling Coil Capacity Information,Coil:Cooling:Water,COIL COOLING WATER 11,26395.21,17356.31,9038.90,0.66,3041.24,30.85
```

Nigusse commented:
Pushed updated (transitioned) defect file to GitHub remote.

Nigusse committed:
adds unit to User Design Capacity header

Nigusse commented:
Added units for the header for `User Design Capacity [W]`. Now the IP units conversion are applied correctly. See below updated IP units in the header and the value also corrected: 

![Branch_System_Sizing_Info2IP2](https://user-images.githubusercontent.com/5805126/234971514-36783527-e89d-4b5d-b757-6316bd3311df.png)

Nigusse commented on src/EnergyPlus/SizingManager.cc:
```diff
@@ -4477,7 +4480,7 @@ void ReportSysSizing(EnergyPlusData &state,
         print(state.files.eio,
               "{}\n",
-              "! <System Sizing Information>, System Name, Load Type, Peak Load Kind, User Design Capacity, Calc Des Air "
+              "! <System Sizing Information>, System Name, Load Type, Peak Load Kind, User Design Capacity [W], Calc Des Air "
```
Added missing units to the `User Design Capacity [W]`. Now the unit appears in the `eio` and `html` table outputs. Without this unit in the table header, units conversion will not be performed correctly if user's preference in `html` tabular output is `InchPound`.

Myoldmopar replied:
Got it.  This looks like something we should refactor later.  Magically hiding the units in the string is unnecessary -- the units should just be an enum passed as a separate argument.  This is reminiscent of the setup output variable refactor.  But that's for a later date.

Nigusse committed:
Merge branch 'develop' into Issue9952_SystemSizingCoolPeak

Nigusse committed:
Updated OutputChanges file

nrel-bot-2 commented:
@Nigusse @myoldmopar it has been 28 days since this pull request was last updated.

Myoldmopar reviewed (approved):
The change from heat to cool timestamp makes sense.  @rraustad or @mjwitte any comments before this merges?

mjwitte commented on src/Transition/OutputRulesFiles/OutputChanges23-1-0-to-23-2-0.md:
```diff
@@ -7,6 +7,10 @@ This file documents the structural changes on the output of EnergyPlus that coul
+### EIO System Sizing Information User Design Capacity header
+Missing unit is added to the EIO ystem Sizing Informationg table "User Design Capacity" header as shown below: 
+- <System Sizing Information>, System Name, Load Type, Peak Load Kind, User Design Capacity [W], Calc Des Air Flow Rate [m3/s], User Des Air Flow Rate [m3/s], Design Day Name, Date/Time of Peak
+- This change also impacts the html tabular output report file
```
My only comment would be to add more information here - report name and sub-table name. Otherwise this looks good.

Nigusse committed:
Merge branch 'develop' into Issue9952_SystemSizingCoolPeak

Nigusse committed:
updated Output Changes file

Myoldmopar commented:
Looks good to me.  Thanks for polishing it up @Nigusse.  Decent shows the expected EIO/Table diffs, but nothing else.  Otherwise it's all happy, and I'm merging this now.

Myoldmopar commented:
I was wrong, this is actually showing some debug failures.  Array index issue in this line:

```
timeStepIndexAtPeakCoolLoad = SysSizPeakDDNum(AirLoopNum).TimeStepAtSensCoolPk(coolPeakDD);
```

where `coolPeakDD` is zero.  @Nigusse do you see an immediate fix for this?  If not, I may revert this PR back out temporarily while we address it and then try the merge again later.

FYI @rraustad @mjwitte

rraustad commented:
Wrap with `if (coolPeakDD > 0)` and then check what happens if `timeStepIndexAtPeakCoolLoad = 0` when reporting.

Nigusse commented:
Which example file or unit test failing? I could test Rich's suggestion. We should not allow to print time stamp if the index to the time is zero. This could be a situation where there is no cooling design period.

Myoldmopar commented:
The 3 EnergyPlusFixture unit tests shown here are failing in debug mode: https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-04fbdf43628806a6566ab41b952db59bb65b17d9-x86_64-Linux-Ubuntu-22.04-gcc-11.3-UnitTestsCoverage-Debug.html#test

Myoldmopar commented:
Adding an `if (coolPeakDD > 0)` in front of that line makes the unit tests pass and shows zero regressions.  Feels like a band-aid, but good to know.  @Nigusse let me know if you can reproduce and push a fix up soon, or I'll push up my hack and we can see what happens.

rraustad commented:
If it is a unit test then I think coolPeakDD would be non-zero during a simulation, even for no cooling load (should probably check that). The unit tests should be updated.

Nigusse commented:
I am going to start testing the unit test now.

Nigusse commented:
I reproduced the defect in debug mode. What is common among these 3 failed unit tests is that all them do not have cooling `SizingPeriod:DesignDay` object:

```
OASystem_HotWaterPreheatCoilScheduledOffSim
OASystem_HotWaterPreheatCoilScheduledOnSim
UnitaryHeatPumpAirToAir_MaxSuppAirTempTest
```
I think the fix should be in `SizingManager`. We do not want this problem to pop up in the future.

Myoldmopar commented:
Thanks @Nigusse, assuming you are pushing up a branch, I'll be ready to review/merge it right away to clean things up.

Nigusse commented:
I ended up creating a new pull request #10071.
